clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,485 @@
+			return nil, err
+		}
+	case "gather":
+		summary, err := npm.gather()
+		if err != nil && !errors.IsMetricViolationError(err) {
+			klog.Error("error gathering metrics: ", err)
+			return nil, err
+		}
+		klog.Info("metric: ", summary)
+		return []measurement.Summary{summary}, err
+	default:
+	}
+	return nil, nil
+}
+
+func (npm *networkPerformanceMeasurement) start(config *measurement.Config) error {
+	npm.initialize(config)
+
+	npm.namespace = netperfNamespace
+	if err := client.CreateNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		return err
+	}
+
+	if err := npm.createClusterRoleBinding(); err != nil {
+		return err
+	}
+
+	if err := npm.createCRD(); err != nil {
+		return err
+	}
+
+	if err := npm.createWorkerPods(); err != nil {
+		return err
+	}
+
+	if err := npm.waitForWorkerPodsReady(); err != nil {
+		return err
+	}
+
+	npm.storeWorkerPods()
+
+	npm.executeTest()
+
+	go npm.watchNetworkTestEvents()
+
+	return nil
+}
+
+func (npm *networkPerformanceMeasurement) executeTest() {
+	switch npm.podRatioType {
+	case oneToOne, manyToMany:
+		npm.execNToMTest()
+	default:
+		klog.Error("Invalid Pod Ratio")
+	}
+}
+
+func (npm *networkPerformanceMeasurement) initialize(config *measurement.Config) {
+	klog.Info("Total configured pods: ", npm.podReplicas)
+	npm.k8sClient = config.ClusterFramework.GetClientSets().GetClient()
+	npm.framework = config.ClusterFramework
+	_, _, npm.podRatioType = deriveClientServerPodNum(npm.podRatio)
+	npm.workerPodList = make(map[string][]workerPodData)
+	npm.podPairCh = make(chan uniquePodPair)
+}
+
+func (npm *networkPerformanceMeasurement) createWorkerPods() error {
+	npm.templateMapping = map[string]interface{}{"Replicas": npm.podReplicas}
+	return npm.framework.ApplyTemplatedManifests(workerPodDeploymentManifestFilePath, npm.templateMapping)
+}
+
+func (npm *networkPerformanceMeasurement) createCRD() error {
+	return npm.framework.ApplyTemplatedManifests(crdManifestFilePath, npm.templateMapping)
+}
+
+func (npm *networkPerformanceMeasurement) deleteCRD() error {
+	gvk := schema.GroupVersionKind{Group: customResourceDefinitionGroup, Kind: customResourceDefinitionKind, Version: "v1"}
+	return npm.framework.DeleteObject(gvk, "", customResourceDefinitionName)
+}
+
+func (npm *networkPerformanceMeasurement) createClusterRoleBinding() error {
+	return npm.framework.ApplyTemplatedManifests(clusterRoleBindingFilePath, npm.templateMapping)
+}
+
+func (npm *networkPerformanceMeasurement) deleteClusterRoleBinding() error {
+	gvk := schema.GroupVersionKind{Group: rbacGroup, Kind: rbacKind, Version: "v1"}
+	return npm.framework.DeleteObject(gvk, "", rbacName)
+}
+
+func (npm *networkPerformanceMeasurement) waitForWorkerPodsReady() error {
+	podReadyPollInterval, weightedPodReadyTimeout := npm.getWeightedTimerValuesForPoll()
+	return wait.Poll(podReadyPollInterval, weightedPodReadyTimeout, npm.checkWorkerPodsReady)
+}
+
+func (npm *networkPerformanceMeasurement) checkWorkerPodsReady() (bool, error) {
```

Comment:
Let's not try to reimplement the logic for waiting for pods becoming ready

you should simply use:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/wait_for_pods.go#L52

Reply:
Logic in wait_for_pods.go doesnot take care of checking whether IPs are allocated to pods. When we tested locally , we saw that pods were UP, but there is a delay in IP allocation. Hence test will eventually fail without IPs. So right now i am not going to use the above utility logic.

Reply:
Wait - that's impossible, because we need to assign IP to even start the pod. I actually can't believe in it...